js/reference.js

```diff
@@ -1114,30 +1114,92 @@ var ERMrest = (function(module) {
+
+                    for (var keyIndex = 0; keyIndex < shortestKeyNames.length; keyIndex++) {
+                        var shortestKey = shortestKeyNames[keyIndex];
+
+                        // shortest key should always be aliased in case that key value was changed
                         // use a suffix of '_o' to represent changes to a value that's in the shortest key that was changed, everything else gets '_n'
-                        submissionData[i][key + oldAlias] = oldData[key];
-                        submissionData[i][key + newAlias] = newData[key];
+                        submissionData[i][shortestKey + oldAlias] = oldData[shortestKey];
+                        submissionData[i][shortestKey + newAlias] = newData[shortestKey];
+
+                        // don't add the current key to the column projections if it's already in there
+                        // this can happen when there are multiple tuples or when a key is part of the visible column set
+                        if (columnProjections.indexOf(shortestKey) === -1) {
+                            // keys are aliased and included in both the keyset and column projections set
+                            columnProjections.push(shortestKey);
+                        }
                     }
-                }
 
-                // The list of column names to use in the uri
-                columnProjections = Object.keys(tuples[0].data);
+                    // Loop through the visible columns so the submission data is based off of the visible columns list
+                    for (var m = 0; m < this.columns.length; m++) {
+                        var column = this.columns[m];
+                        var key;
+
+                        // if the column is disabled (generated or immutable), continue
+                        // no need to submit update data in the submission object
+                        if (column.inputDisabled) {
+                            continue;
+                        }
+
+                        if (column.isPseudo) {
+                            var keyColumns = [];
 
-                // always alias the set of column projections for the key data
-                for (var j = 0; j < columnProjections.length; j++) {
+                            if (column._isKey) {
+                                keyColumns = column.key.colset.columns;
+                            } else if (column._isForeignKey) {
+                                keyColumns =  column.foreignKey.colset.columns;
+                            }
+
+                            for (var n = 0; n < keyColumns.length; n++) {
+                                var referenceColumn = keyColumns[n];
+                                keyColumnName = referenceColumn.name;
+
+                                // here so each column of the columns in keyColumns set is added
+                                if (columnProjections.indexOf(keyColumnName) === -1) {
```

Comment:
These two if statements are duplicate of two if statements after the else. You can create a function to take care of these.

```javascript
var addProjectionAndKeyData = function(index, colName) {
  // add columnProjection
  if (columnProjections.indexOf(colName) === -1) {
    // the list of column names to use in the uri
    columnProjections.push(colName);
  }

  // if the current keyColumnName is in shortestKeyNames, we already added the data to submissionData
  if (shortestKeyNames.indexOf(colName) === -1) {
    // alias all data to prevent aliasing data to the same name as another column that exists in the table
    submissionData[index][colName + newAlias] = newData[colName];
  }
}
```
Then in code
```javascript
if (column.isPseudo) {
  //..
  for () {
    addProjectionAndKeyData(i, keyColumns[n].name);
  }
} else {
  addProjectionAndKeyData(i, column.name);
}
````

You might need to change the function parameters.